ofrak_core/ofrak/model/component_model.py

```diff
@@ -17,6 +20,128 @@ class ComponentConfig:
     """
 
 
+@dataclass
+class ComponentExternalTool:
+    """
+    An external tool or utility (like `zip` or `squashfs`) a component depends on. Includes some
+    basic information on installation, either via package manager or bespoke process.
+
+    There is a set of required package managers: ``ComponentExternalTool.REQUIRED_PKG_MANAGERS``. By
+    default, OFRAK assumes that a tool can be installed with `<pkg_manager> install <tool>`. If this
+    is not the case but the tool can still be installed via package manager (for example, the
+    package containing the tool does not have the same name as the tool itself), custom packages for
+    each manager can be provided. If the tool cannot be installed via one or more package manager
+    at all, the package for that manager should be provided as None, and the `install_hint` must be
+    provided so that a user has some guidance on how to install the tool.
+
+    :ivar tool: Name of the command-line tool that will be run
+    :ivar install_packages: Dictionary of package names to install `tool` for different package
+    managers, including all of ``ComponentExternalTool.REQUIRED_PKG_MANAGERS``; by default, it is
+    assumed that a tool can be installed via a package of the same name, so this argument is only
+    necessary if that is NOT the case for one or more package managers.
+    :ivar install_hints: String to provide guidance for a user on how to install a tool; only
+    required if it is not possible to install via the supported package managers.
+    :ivar install_check_arg: Argument to pass to the tool to check if it can be found and run on
+    the host; defaults to `--help` so that by default, `tool`'s install is checked with:
+        `<tool> --help`
+    """
+
+    tool: str
+    install_packages: Dict[str, Optional[str]] = dataclasses.field(default_factory=dict)
+    install_hints: Optional[str] = None  # e.g. version
+    install_check_arg: str = "--help"
```

Comment:
A few thoughts:
I don't think we should be prescribing any package manager by default. If we're not providing Python infrastructure to install the tool for the user/dev in the first place, then I really don't like it. I also wouldn't set a default `install_check_arg`.

At a minimum, maybe you can consider removing much of this __post_init__ code. I don't think it's necessary.

Continuing...:

Part of me believes folks should still be allowed to craft a shell command inside of a Component implementation complete with flags, piping, other commands despite certain command injection concerns. I concede, I don't really know how truly "dark" a path that is, but this might make the installation of the tool more important to me than wrapping its execution. 

The `Toolchain` implementations handle this differently than what I suggest above, but I propose that our expectations of those programs (compilers, linkers, assemblers) are much more consistent/limited across the board than what we are hoping to do here. (Though that approach may still need to change with enough diversity.)

Installing dependencies for folks also opens a can of worms, but consider the following:
```
AGNUTOOL_TOOL = ExternalTool(
    linux_install="apt install -y agnutool", 
    windows_install="python3 -m pip install \"agnutool @ git+https://github.com/agnutool@master\"", 
    darwin_install="git clone https://github.com/agnutool@master && make -C agnutool && make install"
) 
```

If the external tool is not found on either first attempted usage or through some other preemptive search, maybe OFRAK could prompt the user:
```
$ Could not find <TOOL>, install with <TOOL_INSTALL_COMMAND>? Y/n: n
$ Provide your own install command? Y/n Y
$ sudo install agnutool --dest-path /some/other/place
```

This is closer to what I'm envisioning when we discuss moving away from a soft docker requirement to use OFRAK.

Reply:
I'm curious what problems you see with the default `--help` `install_check` arg?

Reply:
I think I suggested the above because I usually veer away from half measures: Install the thing or don't mention installation at all. Having to maintain Python code that only provides install hints felt like a half measure to me.

I reflected a little bit more on this, the goal of pip install-able OFRAK.

Folks looking for an easy-to-use reverse analysis tool will usually find a git repo with a requirements list, but the various components we have complicate this. After all, why should I have to install the "kitchen sink" when I just want to run this one unpacker?

A requirements list in the form of a Component tree may not be the best alternative. So, here we are, suggestion of what to do if a package isn't found is, after further consideration, a real option here and probably the right one for now. Now, as far as how to suggest...
>I really don't want to try to generalize installation on different systems. 

I propose that `REQUIRED_PKG_MANAGERS: ClassVar[Tuple[str, ...]] = ("brew", "apt")` and everything that uses it is trying to do this. 

I'll revise my feedback to the following:
- I'd rather suggest whole command strings or not suggest anything other than "go install cpio for your platform using your favorite package manager or by building from source." It sounds like, of the two, you prefer the latter. This way we don't need to maintain a required package managers list or validation logic and string formatting for that list...
- To me, there is enough potential variance in a preferred `install_check_arg` that I'd rather it be required _and_ I see future potential for unifying that code to be a version check, deviation from which is a potential source of nasty bugs (unless you want to handle that now). That is where I see a need for some extra validation logic.
- In the future, let's consider providing a prompt for the user to provide the command so they don't have to completely leave the OFRAK runtime (out of scope here I think, but I guess this would be done in the MissingError?)

Reply:
Per offline conversation together and with users, we did decide not to include installing the tools as a direct responsibility of OFRAK in this PR.


There is not longer a default `install_check_arg`, but it is required. For cases where this is not sufficient, the expectation is to subclass `ComponentExternalTool` and override the `is_tool_installed` method with the appropriate check. There are 2 examples of this and documentation to accompany.

There are install command suggestions in the errors raised when a dependency is missing at runtime.